File: src/rac/services/hook.py
"""Git-hook installation — `rac hook install` (v0.13.4).

:func:`install_hook` is the reusable capability behind the CLI adapter: it owns
resource loading, the git-directory check, the never-overwrite refusal, and the
result model. It writes one bundled script into ``<dir>/.git/hooks/<style>``
and marks it executable — note the installed name has no ``.sh`` suffix even
though the packaged resource does. Mirrors ``rac skill install`` (ADR-021
resource loading; never-overwrite posture).

Failure contract:

- unknown style          → :class:`~rac.core.hooks.HookNotFound` (usage error)
- no .git directory      → :class:`NotAGitWorkTree` (usage error, exit 2)
- hook file exists       → :class:`HookFileExists` (refused; exit 1; untouched)
- missing packaged hook  → :class:`~rac.core.hooks.HookResourceMissing`
  (operational; a broken installation)
"""

from __future__ import annotations

import stat
from dataclasses import dataclass
from pathlib import Path

from rac.core.hooks import DEFAULT_STYLE, HookNotFound, available_hooks, load_hook
from rac.errors import RACError

# The executable bits git requires on a hook script: rwxr-xr-x, applied on top
# of whatever mode the freshly written file already carries.
_EXECUTABLE_BITS = stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH


class NotAGitWorkTree(RACError):
    """The target directory has no ``.git`` directory (usage error)."""

    def __init__(self, directory: str):
        self.directory = directory
        super().__init__(
            f"no .git directory in {directory}; run `rac hook install` from a git repository root"
        )


class HookFileExists(RACError):
    """A target hook file already exists; RAC never overwrites it."""

    def __init__(self, path: str):
        self.path = path
        super().__init__(f"{path} already exists; rac hook install never overwrites")


@dataclass
class InstalledHook:
    """Result of a `rac hook install` run (stable JSON contract, ADR-007)."""

    style: str
    path: str
    bytes_written: int

    def to_dict(self) -> dict:
        # bytes_written is tracked for callers but deliberately absent from the
        # JSON shape (pinned by test_installation_json_contract).
        return {
            "schema_version": "1",
            "installed": True,
            "hook": {"style": self.style, "path": self.path},
        }


def install_hook(target_dir: str, style: str = DEFAULT_STYLE) -> InstalledHook:
    """Write the bundled ``style`` hook into ``target_dir``'s ``.git/hooks``.

    Raises :class:`~rac.core.hooks.HookNotFound` for an unknown style,
    :class:`NotAGitWorkTree` when there is no ``.git`` directory,
    :class:`HookFileExists` when the target hook already exists (left
    untouched), and :class:`~rac.core.hooks.HookResourceMissing` when the
    packaged resource is absent.
    """
    if style not in available_hooks():
        raise HookNotFound(style)

    git_dir = Path(target_dir) / ".git"
    if not git_dir.is_dir():
        raise NotAGitWorkTree(target_dir)

    content = load_hook(style)  # cheap, and validates the packaged resource
    dest = git_dir / "hooks" / style
    if dest.exists():
        raise HookFileExists(str(dest))

    dest.parent.mkdir(parents=True, exist_ok=True)
    dest.write_bytes(content)
    dest.chmod(dest.stat().st_mode | _EXECUTABLE_BITS)
    return InstalledHook(style=style, path=str(dest), bytes_written=len(content))

File: rust/rac-engine/src/hook.rs
//! Bundled git hooks — `decided hook` (PORT-CONTRACT.d/15).
//!
//! Port of `src/asdecided/core/hooks.py` (registry + resource loading) and
//! `src/asdecided/services/hook.py` (`install_hook`). The packaged `<style>.sh`
//! scripts are embedded verbatim from `rust/decided-engine/assets/hooks/`,
//! vendored byte-identical copies of the Python package files (unit test
//! below); the INSTALLED file is named `<style>` with no extension, under
//! `<dir>/.git/hooks/`, and is made executable (hook brief, landmines 1-2).

use std::path::Path;

use crate::walk::py_join;

/// One bundled git hook: its style (the git hook filename) and description.
pub struct HookSpec {
    pub style: &'static str,
    pub description: &'static str,
}

/// Bundled hooks, registry order. `install` defaults to the first.
pub const BUNDLED_HOOKS: [HookSpec; 2] = [
    HookSpec {
        style: "post-commit",
        description: "Advisory write-cadence nudge after each commit (never blocks).",
    },
    HookSpec {
        style: "pre-commit",
        description: "Validate staged Markdown artifacts before each commit (blocks on errors).",
    },
];

/// `DEFAULT_STYLE` — the first bundled hook.
pub const DEFAULT_STYLE: &str = "post-commit";

/// The embedded hook scripts, index-aligned with [`BUNDLED_HOOKS`].
pub(crate) const HOOK_BYTES: [&[u8]; 2] = [
    include_bytes!("../assets/hooks/post-commit.sh"),
    include_bytes!("../assets/hooks/pre-commit.sh"),
];

/// `available_hooks()` — bundled hook styles, registry order.
pub fn available_hooks() -> Vec<&'static str> {
    BUNDLED_HOOKS.iter().map(|h| h.style).collect()
}

fn hook_bytes(style: &str) -> Option<&'static [u8]> {
    BUNDLED_HOOKS
        .iter()
        .position(|h| h.style == style)
        .map(|i| HOOK_BYTES[i])
}

/// Result of a `decided hook install` run (`InstalledHook`; `bytes_written` is
/// in the oracle's model but absent from its JSON).
pub struct InstalledHook {
    pub style: String,
    pub path: String,
}

/// The failure contract of `install_hook`, message-shaped like the oracle.
/// `HookNotFound` is unreachable via the CLI (argparse `--style` choices
/// fire first) and is folded into the usage-error path by the caller.
pub enum HookInstallError {
    /// `NotAGitWorkTree` — no `.git` DIRECTORY (a `.git` file — a worktree
    /// or submodule pointer — fails too). Usage error, exit 2.
    NotAGitWorkTree(String),
    /// `HookFileExists` — refused; the existing hook is untouched (exit 1).
    FileExists(String),
    /// Filesystem write failure.
    Io(String),
}

/// `install_hook(target_dir, style)` — write the bundled `style` script to
/// `<dir>/.git/hooks/<style>` and make it executable (git requires the exec
/// bit; the oracle ORs `S_IXUSR|S_IXGRP|S_IXOTH` onto the fresh file's mode,
/// yielding 0755 under the usual umask).
pub fn install_hook(target_dir: &str, style: &str) -> Result<InstalledHook, HookInstallError> {
    let content = hook_bytes(style).expect("argparse-validated style");

    let git_dir = Path::new(target_dir).join(".git");
    if !git_dir.is_dir() {
        return Err(HookInstallError::NotAGitWorkTree(format!(
            "no .git directory in {target_dir}; run `decided hook install` from a git repository root"
        )));
    }

    let dest_display = py_join(target_dir, &[".git", "hooks", style]);
    let dest = Path::new(&dest_display);
    if dest.exists() {
        return Err(HookInstallError::FileExists(format!(
            "{dest_display} already exists; decided hook install never overwrites"
        )));
    }

    let hooks_dir = git_dir.join("hooks");
    std::fs::create_dir_all(&hooks_dir)
        .map_err(|e| HookInstallError::Io(format!("{e}: {}", hooks_dir.display())))?;
    std::fs::write(dest, content)
        .map_err(|e| HookInstallError::Io(format!("{e}: {dest_display}")))?;
    // dest.chmod(dest.stat().st_mode | S_IXUSR | S_IXGRP | S_IXOTH)
    let mode = std::fs::metadata(dest)
        .map_err(|e| HookInstallError::Io(format!("{e}: {dest_display}")))?
        .permissions();
    {
        use std::os::unix::fs::PermissionsExt;
        let new_mode = mode.mode() | 0o111;
        std::fs::set_permissions(dest, std::fs::Permissions::from_mode(new_mode))
            .map_err(|e| HookInstallError::Io(format!("{e}: {dest_display}")))?;
    }
    Ok(InstalledHook {
        style: style.to_string(),
        path: dest_display,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The embedded scripts must be byte-identical to the Python package
    /// files the retirement oracle installs.
    #[test]
    fn embedded_bytes_equal_python_package_files() {
        for (i, spec) in BUNDLED_HOOKS.iter().enumerate() {
            let py_path = format!(
                "{}/../../src/asdecided/hooks/{}.sh",
                env!("CARGO_MANIFEST_DIR"),
                spec.style
            );
            let py_bytes = std::fs::read(&py_path)
                .unwrap_or_else(|e| panic!("cannot read {py_path}: {e}"));
            assert_eq!(
                py_bytes, HOOK_BYTES[i],
                "embedded {} differs from the Python package file",
                spec.style
            );
        }
    }

    #[test]
    fn registry_order_and_default() {
        assert_eq!(available_hooks(), vec!["post-commit", "pre-commit"]);
        assert_eq!(DEFAULT_STYLE, "post-commit");
    }
}
